tests/usb_interfaces.test.cpp

```diff
@@ -0,0 +1,737 @@
+                       std::string_view p_iface_name_two)
+    : m_name_one(p_iface_name_one)
+    , m_name_two(p_iface_name_two) {};
+
+  struct mock_iface_descriptor
+  {
+    byte num;
+    byte alt_settings;
+    byte str_idx;
+    bool feature;
+  };
+
+private:
+  [[nodiscard]] descriptor_delta driver_write_descriptors(
+    endpoint_writer const& p_callback,
+    u8 p_iface_number,
+    u8 p_str_starting_idx) override
+  {
+    std::array<byte const, 8> iad_buf{
+      iad_length,
+      iad_type,
+      0,                    // first interface
+      2,                    // iface count
+      0,                    // class
+      0,                    // subclass
+      0,                    // proto
+      p_str_starting_idx++  // string idx
+    };
+
+    std::array<byte const, 2> iface_header = { iface_desc_length,
+                                               iface_desc_type };
+    std::array<byte const, 5> static_desc_vars = {
+      0,  // altsettings
+      1,  // num endpoints
+      0,  // class
+      0,  // subclass
+      0,  // protocol
+    };
+
+    m_iface_one = { p_iface_number++, 0, p_str_starting_idx++, false };
+    m_iface_two = { p_iface_number++, 0, p_str_starting_idx++, false };
+
+    std::array<byte const, 1> iface_one_arr({ m_iface_one.num });
+    std::array<byte const, 1> iface_one_str_arr({ m_iface_one.str_idx });
+
+    std::array<byte const, 1> iface_two_arr({ m_iface_two.num });
+    std::array<byte const, 1> iface_two_str_arr({ m_iface_two.str_idx });
+
+    auto span_arr = make_scatter_bytes(iad_buf,
+
+                                       // iface one
+                                       iface_header,
+                                       std::span(iface_one_arr),
+                                       static_desc_vars,
+                                       std::span(iface_one_str_arr),
+
+                                       // iface two
+                                       iface_header,
+                                       std::span(iface_two_arr),
+                                       static_desc_vars,
+                                       std::span(iface_two_str_arr));
+
+    p_callback(span_arr);
+    m_wrote_descriptors = true;
+    return { 2, 3 };
+  }
+
+  [[nodiscard]] bool driver_write_string_descriptor(
+    endpoint_writer const& p_callback,
+    u8 p_index) override
+  {
+    if (!m_wrote_descriptors) {
+      // std::println("You must call write_descriptors first");
+      safe_throw(hal::operation_not_permitted(this));
+    }
+    std::array<byte, 2> header{ 0, str_desc_type };
+    if (p_index == m_iface_one.str_idx) {
+      header[0] = m_name_one.length() + 2;
+      auto arr = make_scatter_bytes(
+        header,
+        std::span(reinterpret_cast<byte const*>(m_name_one.data()),
+                  m_name_one.length()));
+      p_callback(arr);
+      return true;
+    } else if (p_index == m_iface_two.str_idx) {
+      header[0] = m_iface_two.str_idx + 2;
```

Comment:
Why does the above `header[0]` get set to the length+2 and this one gets set to str_idx +2?